### ISSUE TITLE
Investigate why save on export triggers on C-c C-c after Org capture

### ISSUE THREAD

**Issue by kaushalmodi** [kaushalmodi/ox-hugo#212]

**Comment by takaxp** [comment]
Hi.

I investigated too. Then, that's because probably the following my setting.

```elisp
(defun my-org-set-created-property (&optional active NAME)
      "Set a property on the entry giving the creation time.

By default the property is called CREATED. If given the `NAME'
argument will be used instead. If the property already exists, it
will not be modified."
      (interactive)
      (let* ((created (or NAME my-org-created-property-name))
             (fmt (if active "<%s>" "[%s]"))
             (now  (format fmt (format-time-string "%Y-%m-%d %a %H:%M"))))
        (unless (org-entry-get (point) created nil)
          (org-set-property created now))))

(add-hook 'org-capture-before-finalize-hook #'my-org-set-created-property)
```

**Comment by takaxp** [comment]
Yes. I'll do that :)

> Also the snippet you pasted above shouldn't affect this because it is not saving the buffer.

I don't think so because the snippet add `:CREATED:  [2018-09-27 Thu 03:27]` in `PROPERTIES` after switching the buffer to my org file for blogging by `C-u C-c C-c` for org-capture. Note that when using `C-c C-c` does not create `2018/.md` but `C-u C-c C-c` creates `2018/.md`.

**Comment by takaxp** [comment]
So, first of all, after upgrading to ox-hugo-20180926.1817, the error has been corrected :)

I'll provide further information soon. (I have to remove shutup.el to show the messages)

**Comment by kaushalmodi** [comment]
> So, first of all, after upgrading to ox-hugo-20180926.1817, the error has been corrected :)

Nice, Melpa picked up that https://github.com/kaushalmodi/ox-hugo/commit/67c1295aa95f21e2e6f800e79c50e75ae91cfe7b commit that was committed just 1/2 hr ago.

> I'll provide further information soon. (I have to remove shutup.el to show the messages)

Also ensure to restart Emacs; especially useful when dealing with hook updates like in that last commit.

**Comment by takaxp** [comment]
Yep. Thanks to MELPA.

Here is my log:
```
This is the last note stored by a capture process
Saving file /Users/taka/Dropbox/org/blog/entries/imadenale.org...
Wrote /Users/taka/Dropbox/org/blog/entries/imadenale.org
[ox-hugo export after save DBG] org-hugo-auto-export-on-save = t
[ox-hugo export after save DBG] org-capture-before-finalize-hook = (org-hugo--disallow-auto-export-during-capture my-org-set-created-property my-org-agenda-to-appt)
[ox-hugo export after save DBG] org-hugo--org-capture-active-flag = nil
imadenale.org: Point is not in a valid Hugo post subtree; move to one and try again
```

And my environment is
1. GNU Emacs 26.1 (build 1, x86_64-apple-darwin17.7.0, NS appkit-1561.60 Version 10.13.6 (Build 17G65)) of 2018-07-20
2. Org mode version 9.1.14 (release_9.1.14-933-gfe72a0.dirty @ /Users/taka/devel/git/org-mode/lisp/)
3. ox-hugo-20180926.1817 from MELPA

I always install Org Mode from the official git server.

Additionally, `my-org-agenda-to-appt` is my private implementation as
```
(defun my-org-agenda-to-appt ()
  (interactive)
  (org-agenda-to-appt t '((headline "TODO"))))
```

Any insights?

**Comment by takaxp** [comment]
The branch/version of Org Mode is master/fe72a05ba.

**Comment by kaushalmodi** [comment]
> The branch/version of Org Mode is master/fe72a05ba.

OK, so you are on a commit with `save-buffer` removed from withing `org-capture-finalize`. 

While debugging this, I went down that rabbit hole and emailed this: https://lists.gnu.org/r/emacs-orgmode/2018-09/msg00289.html :)

**Comment by takaxp** [comment]
```
This is the last note stored by a capture process
Saving file /Users/taka/Dropbox/org/blog/entries/imadenale.org...
Wrote /Users/taka/Dropbox/org/blog/entries/imadenale.org
[ox-hugo export after save DBG] after-save-hook = (ac-clear-variables-after-save t org-hugo-export-wim-to-md-after-save)
[ox-hugo export after save DBG] org-hugo-auto-export-on-save = t
[ox-hugo export after save DBG] org-capture-before-finalize-hook = (org-hugo--disallow-auto-export-during-capture my-org-set-created-property my-org-agenda-to-appt)
[ox-hugo export after save DBG] org-capture-after-finalize-hook = (org-hugo--allow-auto-export-after-capture)
[ox-hugo export after save DBG] org-hugo--org-capture-active-flag = nil
imadenale.org: Point is not in a valid Hugo post subtree; move to one and try again
```

**Comment by kaushalmodi** [comment]
Yes!! That's what I guessed.

See the order of "AFTER capture" and "Saving file ..". They are reversed in your case. So something in your config is saving the imadenale.org **after** the capturing got over.

Find out what it is.

**Comment by takaxp** [comment]
What's happened??? I cannot understand this behavior. LOL.

**Comment by kaushalmodi** [comment]
- `org-hugo-export-wim-to-md-after-save` is added to `after-save-hook` in ox-hugo-auto-export.el.
- The "[ox-hugo export after save DBG]" messages are from that `org-hugo-export-wim-to-md-after-save` fn.
- So each time to save i.e. call `save-buffer` (manually or automatically from some hook or some minor mode), `org-hugo-export-wim-to-md-after-save` gets executed and prints those debug messages.
- Now ox-hugo-auto-export.el configures the `org-capture-before-finalize-hook` and `org-capture-after-finalize-hook` hooks assuming that the saving of capture will happen **between** the calling of the before finalizer hook and after finalizer hook.
- The before finalizer hook sets `org-hugo--org-capture-active-flag` to `t` and when the value of that variable is t, auto export is prevented, as you can see in https://github.com/kaushalmodi/ox-hugo/blob/2be4bc02629bee8848f868fdfc35d9904a82d53a/ox-hugo-auto-export.el#L84-L86.
- And the after finalizer hook sets `org-hugo--org-capture-active-flag` back to nil. So that resumes the "auto export on save" behavior.

From your log:

```
AFTER capture
This is the last note stored by a capture process
Saving file /Users/taka/Dropbox/org/blog/entries/imadenale.org...
```

1. First AFTER capture hook is called, which restores that var back to nil.
2. Then something is saving your buffer. So that's the problem.

**Comment by kaushalmodi** [comment]
> Sorry, I'll leave to sleep for tomorrow. Let's find the solution later again :)

No worries. I hope my last comment helps you figure this out. If not, start by commenting out your whole config (except the ox-hugo stuff) and see what in your config introduces that auto-saving after capture.

**Comment by takaxp** [comment]
Thanks, actually I use some additional settings to save buffer to backup the latest file by `backup-each-save.el`. Thank you for everything and Good night :)

**Comment by kaushalmodi** [comment]
@takaxp Probably the auto-saving thing in your config is the culprit?

https://github.com/takaxp/emacs.d/blob/e59b2aec1627e0f394ec5417f11a8cb42a873d69/init.el#L1810-L1823

**Comment by kaushalmodi** [comment]
Unrelated, but this in your config is also problematic.. https://github.com/takaxp/emacs.d/blob/e59b2aec1627e0f394ec5417f11a8cb42a873d69/init.el#L3311-L3313

You need to have something like:

```el
(require 'ox-hugo-auto-export nil t)
(with-eval-after-load "ox"
  (when (require 'ox-hugo nil t)
```

i.e. don't load `ox-hugo-auto-export` *after* loading `ox-hugo`.. because in a fresh emacs session, when ox-hugo is not already loaded, the auto exporting will auto-load ox-hugo :)

That's why in https://ox-hugo.scripter.co/doc/usage/, I keep the ox-hugo-auto-export outside the ox-hugo form:

```el
(with-eval-after-load 'ox
  (require 'ox-hugo))
(require 'ox-hugo-auto-export)
```

**Comment by takaxp** [comment]
Ouch! I'll fix the require procedure before trying to find the issue on my org-capture. Thanks :)

**Comment by takaxp** [comment]
OK, I changed my setting for `ox-hugo` as follows. `ox-hugo-auto-export.el` is loaded before `ox-hugo.el` (verified).

```
(with-eval-after-load "org"
  (require 'ox-hugo-auto-export nil t))

(with-eval-after-load "ox"
  (when (require 'ox-hugo nil t) ...)
```

And I disabled `auto-save-buffers.el` and `backup-each-save.el`.

After calling `org-capture` and `C-c C-c`, then the log says:

```
Clipboard pasted as level 2 subtree
BEFORE capture
AFTER capture
```

but the imadenale.org has not been saved..., which means `after-save-hook` had not been run during finalizing `org-cpature`.

So I'll try this with minimal settings later today.

**Comment by kaushalmodi** [comment]
> but the imadenale.org has not been saved..., which means after-save-hook had not been run during finalizing org-cpature.

Correct! That's the expected behavior on Org mode devel (built from master branch). Auto-saving after capture got removed in master branch since this commit: https://code.orgmode.org/bzg/org-mode/commit/b4422add3745c26ec3b2e11b8da425844b2e9d3d

I re-enable that auto-saving during capture using this in my config:

```el
(defun modi/org-capture-before-finalize ()
  (let ((base-buffer (buffer-base-buffer (current-buffer))))
    (org-capture-put :base-buffer base-buffer)))
(add-hook 'org-capture-before-finalize-hook #'modi/org-capture-before-finalize)

(defun modi/org-capture-save-buffer ()
  (when-let ((base-buffer (org-capture-get :base-buffer)))
    (with-current-buffer base-buffer
      (save-buffer))))
(add-hook 'org-capture-after-finalize-hook #'modi/org-capture-save-buffer)
```

**Comment by takaxp** [comment]
OK, now I understand what is the reason of this issue and find a solution. The solution is a combination of using your nice functions `modi/org-capture-get-base-buffer` and `modi/org-capture-save-base-buffer`, and applying some pre-processing for `auto-save-buffer.el` so that the auto-saving by the elisp will not activated when `org-capture-mode` is `t`.

After adopting the combination, the log says finally as follows:
```
Clipboard pasted as level 2 subtree
BEFORE capture
1 bookmark(s) restored based on context.
Added 9 events for today
Saving file /Users/taka/Dropbox/org/blog/entries/imadenale.org...
Wrote /Users/taka/Dropbox/org/blog/entries/imadenale.org
[ox-hugo export after save DBG] org-hugo-auto-export-on-save = t
[ox-hugo export after save DBG] org-capture-before-finalize-hook = (modi/org-capture-get-base-buffer org-hugo--disallow-auto-export-during-capture my-org-set-created-property my-org-agenda-to-appt)
[ox-hugo export after save DBG] org-hugo--org-capture-active-flag = t
AFTER capture
```

I'd like to wrap up this issue by reporting what I verified.

The problem consists of the following elements.
1. The latest Org Mode does not auto-saving during org-capture, and I didn't use your nice auto-saving functions for org-capture yet
2. I use `auto-save-buffers.el`
3. I use `my-org-set-created-property`

The condition of `1`+`2`+`3` causes the error that we looked double auto-saving and unexpected behavior observed as a reversed capturing order.

The double auto-saving has two stages. The fist stage is happened before logging "AFTER capture" because `auto-save-buffers.el` saves the buffer due to the above condition `1`. The second stage is happened after logging "BEFORE capture" also because `auto-save-buffer.el` saves the buffer again due to `my-org-set-created-property` added `CREATED` property (condition `3`).

Once I introduced your nice auto-saving functions, the second stage did not happen. In stead of that, another auto-saving was logged before "AFTER capture" as you expected but the first stage still exists.

The first stage auto-saving by `auto-save-buffers.el` is actually triggered by calling org-capture, but I found the template buffer of org-capture uses a minor mode of `org-capture-mode`. So I added an additional setting to ignore the first auto-saving stage by checking the minor mode is active or not. Then finally I reached the solution of this issue... Woooo!

It is really really complicated and I couldn't find the solution by myself but I did by your kindly supports.

I really appreciate your help! Thanks!
